### PR TITLE
EDU-2785 - Create Audiobooks with OpenAI

### PR DESCRIPTION
**Pull request by fairlydurable**
Ready for review

### CODE REVIEW

File: docs/getting_started/python/first_program_in_python/index.md [hunk 1]
```diff
@@ -218,7 +219,7 @@ First, the `withdraw()` Activity takes the details about the transfer and calls
         reference_id = f"{data.reference_id}-withdrawal"
         try:
             confirmation = await asyncio.to_thread(
-                self.bank.withdraw, data.source_account, data.amount, 
+                self.bank.withdraw, data.source_account, data.amount, reference_id
```
Comment (jsundai) [inline docs/getting_started/python/first_program_in_python/index.md:222]: Did this maybe slip into the commit?

File: docs/tutorials/java/audiobook/index.md [hunk 2]
```diff
@@ -0,0 +1,982 @@
...
+
+## Introduction
+
+For many people, audiobooks are a staple of daily life.
+Spoken content helps people pass time, learn new things, and stay entertained without tying themselves to a screen or a book.
+Imagine turning your daily commute, workout sessions, or household chores into opportunities to catch up on _all_ your reading -- not just books.
+You can grab text from emails, from the web, or even that report you need to "read" for work.
+With the rise of advanced text-to-speech (TTS) technology, it's not just possible, it's quick to build in Java with simple APIs.
+
+For years, text-to-speech solutions supported the assistive community by transforming words on the page into accessible engagement.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:23]:
```suggestion
For years, text-to-speech solutions have supported the assistive community by transforming words on the page into accessible engagement.
```

File: docs/tutorials/java/audiobook/index.md [hunk 3]
```diff
@@ -0,0 +1,982 @@
...
+However, until recently, robotic voices and unnatural rhythms made TTS unappealing for casual listeners.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:24]: This intro is a bit long IMO. I might take out the paragraph starting at line 24-25 and shorten this to 2-3 paragraphs.

File: docs/tutorials/java/audiobook/index.md [hunk 4]
```diff
@@ -0,0 +1,982 @@
...
+Using these large language models (LLMs), you can enjoy warm, nuanced narration.
+So, wouldn't it be great to use this kind of tech to transform any text into an audiobook?
+
+In this tutorial, you'll use LLM voices to transform any text into audiobooks.
+You'll build a robust text-to-speech system with OpenAI APIs and Temporal Technology's error mitigation.
+Forget about manually checking statuses or handling errors.
+Just focus on converting text while Temporal makes sure everything runs smoothly.
+When complete, you'll be able to reliably convert any plain text file into an audiobook.
+
+Ready to transform your text into immersive audiobooks?
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:38]: These two sentences might not be needed.

File: docs/tutorials/java/audiobook/index.md [hunk 5]
```diff
@@ -0,0 +1,982 @@
...
+1. **Set up your project directory**:
+   Set up your source code folder hierarchy by issuing this directory creation command:
+
+   ```sh
+   mkdir -p TTSWorker/src/main/java/ttsworker
+   ```
+
+   Your directory structure should now look like this:
+
+   ```text
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:84]: love this

File: docs/tutorials/java/audiobook/index.md [hunk 6]
```diff
@@ -0,0 +1,982 @@
...
+:::
+
+## Create your OpenAI conversion code
+
+Create an interface-implementation pair of files named `TTSActivities.java` and `TTSActivitiesImpl.java`.
+Add these to 'src/main/java/ttsworker' and include the following file contents.
+This class is responsible for the text-to-speech work in your project:
+
+<details>
+
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:175]: The Hover sentence could be made into a note or something similar. "Hover your cursor over the code block to reveal the copy-code option."

File: docs/tutorials/java/audiobook/index.md [hunk 7]
```diff
@@ -0,0 +1,982 @@
...
+}
+```
+<!--SNIPEND-->
+
+  </TabItem>
+</Tabs>
+
+</details>
+
+Activities, like the `TTSActivities` class, handle potentially unreliable parts of your code, such as calling APIs or working with file systems.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:384]: Could add a link to activities in the encyclopedia.

File: docs/tutorials/java/audiobook/index.md [hunk 8]
```diff
@@ -0,0 +1,982 @@
...
+Temporal uses Activities for any action that might have to be retried.
+Imagine that the network goes down or your service provider (OpenAI in this case) is temporarily doing maintenance.
+Activities add check-in points for your application state, which are maintained in the Temporal system and become part of your conversion history.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:387]: I'm not sure about this sentence in terms of activities adding 'check-in points' in a general sense. I might rework it.

File: docs/tutorials/java/audiobook/index.md [hunk 9]
```diff
@@ -0,0 +1,982 @@
...
+Activities let you try your call again in a few seconds, a few minutes, or even later.
+The work automatically picks up
+
+The `TTSActivities` class reads strings from files and converts those strings to audio.
+The class has four activities:
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:392]:
```suggestion
This class has four activities:
```

**Review by jsundai — COMMENTED** [review]
Fantastic work! Went through the beginning, will pick up the rest when available.

File: docs/getting_started/python/first_program_in_python/index.md (hunk 1 above)
Comment (fairlydurable) [reply docs/getting_started/python/first_program_in_python/index.md:222]: It almost certainly did. I'll remove it from the commit.

File: docs/tutorials/java/audiobook/index.md (hunk 3 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:24]: I talked to Brian about this. I really want to acknowledge the assistive community.

File: docs/tutorials/java/audiobook/index.md (hunk 4 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:38]: They're part of the rubric, that we need to have transitions. Thank you for the check though.

File: docs/tutorials/java/audiobook/index.md (hunk 6 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:175]: You don't see the hover sentence unless you do the toggle, so it's part of that infrastructure. If you download and build, it will be more obvious.

Thanks for the check in.

File: docs/tutorials/java/audiobook/index.md (hunk 7 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:384]: As I should. Adding!

File: docs/tutorials/java/audiobook/index.md (hunk 8 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:387]: Thanks for pointing that out. I've noted it down for further discussion with people / potential rewriting.

File: docs/tutorials/java/audiobook/index.md (hunk 9 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:392]: I'm going to add this directly on my side, but the change is adopted. Thanks!

File: docs/tutorials/java/audiobook/index.md (hunk 3 above)
Comment (jsundai) [reply docs/tutorials/java/audiobook/index.md:24]: Ok cool. I might still polish the intro a bit and either shorten it or focus it a bit more if you can. I like how in Patrick's saga patten - it talks about saga pattern in general, the saga pattern in this application, why temporal, then the transition.

File: docs/tutorials/java/audiobook/index.md [hunk 10]
```diff
@@ -0,0 +1,982 @@
...
+
+When using your application, you must set your `OPEN_AI_BEARER_TOKEN` environment variable in the same shell before execution.
+The application checks for the token and if it's not set, it will error.
+
+:::
+
+## Create your OpenAI conversion code
+
+Create an interface-implementation pair of files named `TTSActivities.java` and `TTSActivitiesImpl.java`.
+Add these to 'src/main/java/ttsworker' and include the following file contents.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:171]: 'src/main/java/ttsworker' -> `src/main/java/ttsworker`

File: docs/tutorials/java/audiobook/index.md [hunk 11]
```diff
@@ -0,0 +1,982 @@
...
+
+  </TabItem>
+</Tabs>
+
+</details>
+
+[Activities](https://docs.temporal.io/activities), like the `TTSActivities` class, handle potentially unreliable parts of your code, such as calling APIs or working with file systems.
+Temporal uses Activities for any action that might have to be retried.
+Imagine that the network goes down or your service provider (OpenAI in this case) is temporarily doing maintenance.
+Your retry policy lets you perform "do-over"s.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:387]: small nit - should it be "do-overs"?

File: docs/tutorials/java/audiobook/index.md [hunk 12]
```diff
@@ -0,0 +1,982 @@
...
+- **`moveOutputFileToPlace`**: Find a safe, versioned location for the audio in the same folder as the original text, and move it there.
+
+Here is how the activities help in the overall conversion process:
+![After reading a text file and dividing it into chunks, each chunk is sent to OpenAI to be converted to audio and the results appended to the output file](images/highlevelprocess.png)
+
+### Prepare your text with `readFile`
+
+The `readfile` Activity reads and processes text from a file.
+Before and during the read, it performs a series of safety checks, such as making sure the file exists and it's readable.
+If it encounters any file system error, the project assumes the situation is unrecoverable, throwing a non-retryable exception, ending the conversion process.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:406]:
```suggestion
If it encounters any file system error, the project assumes the situation is unrecoverable, throwing a non-retryable exception, and ending the conversion process.
```

File: docs/tutorials/java/audiobook/index.md [hunk 13]
```diff
@@ -0,0 +1,982 @@
...
+        return outputPath;
+    }
+}
+
+```
+
+This class creates `ActivityOptions`, which set the policies Temporal uses for retrying failed Activities.
+Then it builds the `encodingStub`, which allows your application to run Activities as they are managed by the Temporal system.
+Temporal management (called "orchestration") enables the Temporal Service to track Workflow and Activity progress and state, and to manage Activities when they encounter errors.
+This progress is stored centrally on the Temporal Server, so it can resume if interrupted.
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:744]: I would break up this paragraph

File: docs/tutorials/java/audiobook/index.md [hunk 14]
```diff
@@ -0,0 +1,982 @@
...
+    result                        Ok
+```
+
+## Wrap-up
+
+In this tutorial, you created an OpenAI solution that converts text files into audio.
+You used Temporal error mitigation to make sure that failed API requests could be retried and catastrophic events could be recovered.
+With just a few source files, you created a complete working solution to build a durable, reliable system to build audiobooks from simple text files.
+
+Now that you've completed this tutorial, check out some other great  [Temporal Java projects](https://learn.temporal.io/tutorials/java/).
```
Comment (jsundai) [inline docs/tutorials/java/audiobook/index.md:981]: could make this an admonition?

**Review by jsundai — COMMENTED** [review]
Great work again! Did a prose review.

File: docs/tutorials/java/audiobook/index.md (hunk 10 above)
Comment (jsundai) [reply docs/tutorials/java/audiobook/index.md:171]: needs the backticks

**Review by MasonEgger — DISMISSED** [review]
First pass. Great job!

File: docs/tutorials/java/audiobook/index.md (hunk 3 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:24]: Reduced paragraph count.

File: docs/tutorials/java/audiobook/index.md (hunk 10 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:171]: Single quotes for paths all updated to backticks. I checked various standards to confirm.

File: docs/tutorials/java/audiobook/index.md (hunk 11 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:387]: Updated as recced

File: docs/tutorials/java/audiobook/index.md (hunk 12 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:406]: adopted

File: docs/tutorials/java/audiobook/index.md (hunk 13 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:744]: broken.

File: docs/tutorials/java/audiobook/index.md [hunk 15]
```diff
@@ -0,0 +1,982 @@
...
+Spoken content helps people pass time, learn new things, and stay entertained without tying themselves to a screen or a book.
+Imagine turning your daily commute, workout sessions, or household chores into opportunities to catch up on _all_ your reading -- not just books.
+You can grab text from emails, from the web, or even that report you need to "read" for work.
+With the rise of advanced text-to-speech (TTS) technology, it's not just possible, it's quick to build in Java with simple APIs.
+
+For years, text-to-speech solutions have supported the assistive community by transforming words on the page into accessible engagement.
+However, until recently, robotic voices and unnatural rhythms made TTS unappealing for casual listeners.
+If you didn't need TTS, you probably wouldn't have used TTS.
+
+Now, companies like [OpenAI](https://openai.com/api/) revolutionized speech synthesis.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:27]: adopted

File: docs/tutorials/java/audiobook/index.md [hunk 16]
```diff
@@ -0,0 +1,982 @@
...
+## Prerequisites
+
+You can build the project by following this tutorial or, if you're feeling impatient, just grab the ready-to-go source from its [GitHub repo](https://github.com/temporalio/build-audiobook-java).
+This repository has the full source code and can serve as a guide or a fallback if you encounter issues when working through this tutorial.
+If you want to play first and explore later, you can come back and read through the how-to and background.
+Here's what you need to get going:
+
+1. **A local Java development environment**:
+   Follow ["Set up a local development environment"](/getting_started/java/dev_environment/index.md) so you're ready to build Temporal applications with Java.
+   Ensure a local Temporal Service is running and that you can access the Temporal Web UI from port `8233`.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:50]: I had it originally at 8080 but apparently the other tutorials are using 8233. I'll ask in-channel.

File: docs/tutorials/java/audiobook/index.md [hunk 17]
```diff
@@ -0,0 +1,982 @@
...
+1. **OpenAI API access**:
+   Sign up for your [OpenAI API developer account](https://openai.com/api/) and create a bearer token.
+   You need this token to access OpenAI services for text-to-speech conversion in your project.
+
+If you're all caught up with prerequisites, it's time to build out your Java project directory.
+
+:::info CAUTIONS
+
+This is a tutorial project and its implementation is suited for personal and hobbyist use.
+In production, you wouldn't read or write from a single hard drive.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:66]: I'll adapt to cover both

File: docs/tutorials/java/audiobook/index.md [hunk 18]
```diff
@@ -0,0 +1,982 @@
...
+   ```text
+   TTSWorker
+   └── src
+       └── main
+           └── java
+               └── ttsworker
+   ```
+
+   You can check it with the Unix `tree` command.
+   If the command isn't native to your system, you can install it with standard package managers like Homebrew and apt-get.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:93]: I'm stunned. Adopted.

File: docs/tutorials/java/audiobook/index.md [hunk 19]
```diff
@@ -0,0 +1,982 @@
...
+    mainClass = 'ttspackage.TTSWorkerApp'
+    standardOutput = System.out
+}
+```
+<!--SNIPEND-->
+
+4. **Review your dependencies and run task**:
+   Your dependencies include Temporal's [Java SDK](https://github.com/temporalio/sdk-java), and a few basic libraries:
+
+   * com.squareup.okhttp3:okhttp:4.9.3: OkHttp is a basic HTTP client for network requests.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:139]: you **got** it. adoped

File: docs/tutorials/java/audiobook/index.md [hunk 20]
```diff
@@ -0,0 +1,982 @@
...
+    ID3V2                         no
+    APEV1                         no
+    APEV2                         no
+    last frame
+        offset                    11386560 b (0xadbec0)
+        length                    480
+    errors                        none
+    result                        Ok
+```
+
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:974]: Adding.

File: docs/tutorials/java/audiobook/index.md [hunk 21]
```diff
@@ -0,0 +1,982 @@
...
+   * com.squareup.okhttp3:okhttp:4.9.3: OkHttp is a basic HTTP client for network requests.
+   * org.json:json:20210307: Parse and manipulate JSON data.
+   * commons-io:commons-io:2.11.0: Perform file tasks with common input/output routines.
+   * org.slf4j:slf4j-nop:2.0.6: Minimizes unnecessary output with logging suppression.
+   * io.temporal:temporal-sdk:1.22.2: Add error mitigation.
+
+   The run task starts your TTS application.
+
+5. **Optionally, create a `bearer.sh` utility**:
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:147]: Updated to make this clearer and move the optional down.

File: docs/tutorials/java/audiobook/index.md [hunk 22]
```diff
@@ -0,0 +1,982 @@
...
+   Make it executable with `chmod +x`.
+   Store this wherever you keep similar secure items:
+
+   ```sh
+   #! /bin/sh
+
+   setenv OPEN_AI_BEARER_TOKEN 'your-secret-bearer-token'
+   ```
+
+   This 'bearer.sh' script sets your OpenAI bearer token as an environment variable.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:157]: adopted

File: docs/tutorials/java/audiobook/index.md [hunk 23]
```diff
@@ -0,0 +1,982 @@
...
+:::note Your Bearer Token
+
+When using your application, you must set your `OPEN_AI_BEARER_TOKEN` environment variable in the same shell before execution.
+The application checks for the token and if it's not set, it will error.
+
+:::
+
+## Create your OpenAI conversion code
+
+Create an interface-implementation pair of files named `TTSActivities.java` and `TTSActivitiesImpl.java`.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:170]: adopted

File: docs/tutorials/java/audiobook/index.md [hunk 24]
```diff
@@ -0,0 +1,982 @@
...
+```
+<!--SNIPEND-->
+
+  </TabItem>
+</Tabs>
+
+</details>
+
+[Activities](https://docs.temporal.io/activities), like the `TTSActivities` class, handle potentially unreliable parts of your code, such as calling APIs or working with file systems.
+Temporal uses Activities for any action that might have to be retried.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:385]: adopted

File: docs/tutorials/java/audiobook/index.md (hunk 11 above)
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:387]: Adopted

[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:387]: p.s. that

File: docs/tutorials/java/audiobook/index.md [hunk 25]
```diff
@@ -0,0 +1,982 @@
...
+Imagine that the network goes down or your service provider (OpenAI in this case) is temporarily doing maintenance.
+Your retry policy lets you perform "do-over"s.
+Activities add check-in points for your application state in your [Event History](https://docs.temporal.io/workflows#event-history).
+When retried, it picks up state from just before it was first called, so it's like the failed attempt never happened.
+
+The `TTSActivities` class reads strings from files and converts those strings to audio.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:391]: Like!

File: docs/tutorials/java/audiobook/index.md [hunk 26]
```diff
@@ -0,0 +1,982 @@
...
+This class has four activities:
+
+- **`readFile`**: Pass an `inputPath` to a file, and it returns a chunked list of the contents for well-sized API calls.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:394]: Like. I made them all parallel.

File: docs/tutorials/java/audiobook/index.md [hunk 27]
```diff
@@ -0,0 +1,982 @@
...
+- **`createTemporaryFile`**:  Ask the system to build a temporary file for audio output in a safe folder that normally cleans itself on reboots.
+- **`process`**: Send a text chunk to OpenAI, retrieve the TTS audio segment, and append it to the output file.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:396]: adopted

File: docs/tutorials/java/audiobook/index.md [hunk 28]
```diff
@@ -0,0 +1,982 @@
...
+Imagine that the network goes down or your service provider (OpenAI in this case) is temporarily doing maintenance.
+Your retry policy lets you perform "do-over"s.
+Activities add check-in points for your application state in your [Event History](https://docs.temporal.io/workflows#event-history).
+When retried, it picks up state from just before it was first called, so it's like the failed attempt never happened.
+
+The `TTSActivities` class reads strings from files and converts those strings to audio.
+This class has four activities:
+
+- **`readFile`**: Pass an `inputPath` to a file, and it returns a chunked list of the contents for well-sized API calls.
+- **`createTemporaryFile`**:  Ask the system to build a temporary file for audio output in a safe folder that normally cleans itself on reboots.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:395]: adopted

File: docs/tutorials/java/audiobook/index.md [hunk 29]
```diff
@@ -0,0 +1,982 @@
...
+```
+
+As a rule-of-thumb, a typical English word uses 1.33 OpenAI tokens, which is how this code defines the `AVERAGE_TOKENS_PER_WORD` constant.
+When working with other languages, adjust that value to fit with typical word lengths.
+
+Tokens quantify the data processed by OpenAI requests and all OpenAI endpoints use token limits.
+This code creates chunks with approximately 512 token for each API request.
+Although the OpenAI token limit is higher than this, this code is conservative to reduce bandwidth for the returned audio bytes.
+
+After creating string data for your request stream, build an output file in the next steps.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:444]: Fixing.

File: docs/tutorials/java/audiobook/index.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:50]: 8080

File: docs/tutorials/java/audiobook/index.md [hunk 30]
```diff
@@ -0,0 +1,982 @@
...
+Temporary folders and files avoid name collisions, improve security, and isolate partial work.
+On some systems, temporary folders are optimized for performance so you benefit from faster read and write operations.
+
+With data and an output file, you can begin to create and store your audio.
+
+### Convert text to audio with `process`
+
+The `process` Activity handles text-to-speech work.
+Each time it's called, it sends a chunk of text to the `textToSpeech` method and appends those results to your output file.
+You call it with a string to process and the output destination.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:473]: I struggled with this. OpenAI is on your side. Capitalized and code faced.

File: docs/tutorials/java/audiobook/index.md [hunk 31]
```diff
@@ -0,0 +1,982 @@
...
+The path helps you determine where to move your completed output file after your conversion is done:
+
+```java
+public class TTSActivitiesImpl implements TTSActivities {
+    public static String bearerToken = null;
+    Path canonicalPath = null;
+...
+```
+
+The `textToSpeech` method calls out to OpenAI to convert a `String` into a `byte[]` array.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:507]: I say stet. What kind of array? `byte[]`. Reasoning:

1. into a `byte[]`.
WTF? This reads like you got hit by truck-kun in the middle of saying something

2. into a `byte` array.
Wrong type.

File: docs/tutorials/java/audiobook/index.md [hunk 32]
```diff
@@ -0,0 +1,982 @@
...
+When creating the HTTP request body, customizable [endpoint options](https://platform.openai.com/docs/api-reference/audio/createSpeech) shape the way your audio gets built:
+
+- **model** (required): You'll use the basic low-latency 'tts-1' model in this project.
+  Visit [OpenAI TTS](https://platform.openai.com/docs/models/tts) to read about its available models, which include both standard and high quality options.
+- **input** (required): The maximum length of this string is set at 4096 characters.
+- **voice** (required): The 'nova' voice has a high energy "lively tone".
+  You can listen to samples of other voices at the [OpenAI Voice Options](https://platform.openai.com/docs/guides/text-to-speech/voice-options) page.
+- **response_format**: You'll use the highly portable mp3 output format.
+
+You may want to tweak the request body further.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:550]: adopted but (ahem) tweaked.

File: docs/tutorials/java/audiobook/index.md [hunk 33]
```diff
@@ -0,0 +1,982 @@
...
+
+This section talked about Activities, the individual and potentially unreliable processes that perform work that might fail.
+Next, you'll create a Workflow, which sets the overall business logic for your application.
+
+## Define your Workflow with Temporal
+
+You've seen the pieces that perform the conversion work but you haven't tied them together.
+In this section, you'll build a Temporal Workflow to process a text-to-speech conversion from start to finish.
+[Workflows](https://docs.temporal.io/workflows) create the overall flow of your application's business process.
+It's basically a sequence of steps written in your programming language.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:596]: You convinced me.

File: docs/tutorials/java/audiobook/index.md [hunk 34]
```diff
@@ -0,0 +1,982 @@
...
+
+Create an interface-implementation pair of files named `TTSWorkflow.java` and `TTSWorkflowImpl.java`.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:598]: Agree on sequencers, but not happy with First. Adopting the essence, but slightly rewording.

File: docs/tutorials/java/audiobook/index.md [hunk 35]
```diff
@@ -0,0 +1,982 @@
...
+Add these to 'src/main/java/ttsworker' and include the following file contents.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:599]: Sequencing!

File: docs/tutorials/java/audiobook/index.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:744]: Adopted

File: docs/tutorials/java/audiobook/index.md [hunk 36]
```diff
@@ -0,0 +1,982 @@
...
+It allows your application to mitigate errors and keep going without repeating work you've already done.
+That makes your apps both reliable and efficient.
+
+After the options and encoding stub are two further methods:
+
+- **`fetchMessage`**:
+  This is a Temporal Query method.
+  Some conversion tasks take a while to complete.
+  Queries can "peek" at running processes and fetch information about how that work is proceeding.
+  This method returns a shared `message`, which is updated at each stage of the conversion process.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:756]: Reworded. Thank you.

File: docs/tutorials/java/audiobook/index.md [hunk 37]
```diff
@@ -0,0 +1,982 @@
...
+- **`startWorkflow`**:
+  Every Workflow has one entry point, which is marked by the `@WorkflowMethod` annotation in the Workflow interface.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:758]: Reworded but no doubles on the annotated

File: docs/tutorials/java/audiobook/index.md [hunk 38]
```diff
@@ -0,0 +1,982 @@
...
+  It defines a complete business logic flow, in this case transforming the text within a file pointed to by the `fileInputPath` parameter, into spoken audio.
+  As you see, this code reads the source file, creates the temporary file, processes each chunk, and finally moves the output file into place.
+  The method returns another string, a file output path for the generated mp3 results.
+
+With your Workflows and Activities in place, you can now write a [Worker application](https://docs.temporal.io/workers#worker-program).
+A [Worker](https://docs.temporal.io/workers) hosts your Activities and Workflows and polls a Service-provided Task Queue to look for work to do.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:764]: adopted

File: docs/tutorials/java/audiobook/index.md [hunk 39]
```diff
@@ -0,0 +1,982 @@
...
+
+## Build your Worker
+
+Create `TTSWorkerApp.java` in 'src/main/java/ttsworker' and add the following file contents:
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:768]: Fixed globally.

File: docs/tutorials/java/audiobook/index.md [hunk 40]
```diff
@@ -0,0 +1,982 @@
...
+    }
+}
+```
+<!--SNIPEND-->
+
+Every Worker polls [Task Queues](https://docs.temporal.io/workers#task-queue) to pick up Workflow and Activity Tasks.
+The Task Queue for this project is named `tts-task-queue`.
+You'll use this queue to submit conversion requests.
+Normally you'd build a dedicated full stack solution that submits requests to the Temporal Service and retrieves the results when they're complete.
+For this project, you'll use command-line calls.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:815]: Reworded

File: docs/tutorials/java/audiobook/index.md [hunk 41]
```diff
@@ -0,0 +1,982 @@
...
+
+Some more about his code:
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:817]: Much better. I wonder who "he" was.

File: docs/tutorials/java/audiobook/index.md [hunk 42]
```diff
@@ -0,0 +1,982 @@
...
+
+Make sure the Temporal development server is running and using a persistent store.
+Interrupted work can be picked up and continued without repeating steps, even if you experience server interruptions:
+
+```sh
+temporal server start-dev \
+    --db-filename /path/to/your/temporal.db
+```
+
+Once running, connect to the [Temporal Web UI](http://localhost:8233/) and verify that the server is working.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:859]: Updated globally in file.

File: docs/tutorials/java/audiobook/index.md [hunk 43]
```diff
@@ -0,0 +1,982 @@
...
+```
+
+### Find your output file
+
+Your output is collected in a system-provided temporary file.
+After, your generated MP3 audio is moved into the same folder as your input text file.
+It uses the same name replacing the `txt` extension with `mp3`.
+If an output file already exists, the project versions it to prevent name collisions.
+
+The `TTSWorkflow` returns a string, the /path/to/your/output/audio-file.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:924]: Adopted and audited to find other potential places, anywhere /path/to was used. Found two that I hadn't hit.

File: docs/tutorials/java/audiobook/index.md [hunk 44]
```diff
@@ -0,0 +1,982 @@
...
+Check the Web UI Input and Results section after the Workflow completes.
+The results path is also displayed as part of the CLI's `workflow execute` command output and in the Worker logs.
+
+:::tip Cautions and notes
+
+- Do not modify your input or output files while the Workflow is running.
+  Temporal isn't a cure for bad judgement.
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:931]: Agreed. Sad. But agreed.

File: docs/tutorials/java/audiobook/index.md [hunk 45]
```diff
@@ -0,0 +1,982 @@
...
+    APEV1                         no
+    APEV2                         no
+    last frame
+        offset                    11386560 b (0xadbec0)
+        length                    480
+    errors                        none
+    result                        Ok
+```
+
+## Wrap-up
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:975]: Adopted

File: docs/tutorials/java/audiobook/index.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:981]: Nice.

File: docs/tutorials/java/audiobook/index.md [hunk 46]
```diff
@@ -0,0 +1,982 @@
...
+   └── src
+       └── main
+           └── java
+               └── ttsworker
+   ```
+
+   You can check it with the Unix `tree` command.
+   If the command isn't native to your system, you can install it with standard package managers like Homebrew and apt-get.
+
+2. **Initialize version control**:
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:95]: It gave me an "in" to compare Temporal to version control later...

File: docs/tutorials/java/audiobook/index.md [hunk 47]
```diff
@@ -0,0 +1,982 @@
...
+In production, you wouldn't read or write from a single hard drive.
+A single disk isn't durable so you can't develop durable software with it.
+You must be able to rebuild your progress state or store information somewhere more durable.
+If you expand on this project, consider a API-based Cloud storage solution.
+
+:::
+
+## Create your Java project structure
+
+1. **Set up your project directory**:
```
[the remark this answers is not in the document]
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md:75]: Removed bolding from the start of the numbered lists.

**Comment by fairlydurable** [comment]
Thank you @jsundai and @MasonEgger! Fantastic feedback

File: docs/tutorials/java/audiobook/index.md [hunk 48]
```diff
@@ -0,0 +1,1005 @@
...
+
+   The run task starts your TTS application.
+
+4. Establish your bearer token environment variable
+   Your application uses the environment variable to authenticate with the OpenAI service.
+   It needs to live in the same shell as  your running app.
+
+   This varies by shell, so you might use the following for tcsh:
+
+   ```sh
```
Comment (MasonEgger) [inline docs/tutorials/java/audiobook/index.md:140]: I would swap this with bash. Bash will be more common, so have this be the 'or if you're using a different shell, like tsch'

**Review by MasonEgger — APPROVED** [review]

File: docs/tutorials/java/audiobook/index.md (hunk 48 above)
Comment (fairlydurable) [reply docs/tutorials/java/audiobook/index.md]: tcsh has been wiped from reality. Resolved.

**Comment by fairlydurable** [comment]
* External code review complete. (Thanks @Quinn-With-Two-Ns )
* Code updated.
* Code tested.
* Reviewed tutorial re-snipsynced to bring it inline with updated repo
* Main merged, pushed to GitHub.

Ready to publish
